otherlibs/runtime_events/runtime_events_consumer.c

```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
+    ret = E_MAP_FAILURE;
+    goto failed2;
   }
 
   cursor->ring_file_size_bytes = GetFileSize(cursor->ring_file_handle, NULL);
-#else
-  ring_fd = open(runtime_events_loc, O_RDONLY, 0);
+  return E_SUCCESS;
 
-  if( ring_fd == -1 ) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_OPEN_FAILURE;
-  }
+ failed2:
+  CloseHandle(cursor->ring_handle);
+ failed1:
+  CloseHandle(cursor->ring_file_handle);
+  return ret;
+#else
+  int ring_fd = open(runtime_events_loc, O_RDONLY, 0);
+  if(ring_fd == -1) return E_OPEN_FAILURE;
 
+  struct stat tmp_stat;
   ret = fstat(ring_fd, &tmp_stat);
-
   if (ret < 0) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_OPEN_FAILURE;
+    ret = E_OPEN_FAILURE;
+    goto failed;
   }
-
   cursor->ring_file_size_bytes = tmp_stat.st_size;
 
   /* This cast is necessary for compatibility with Illumos' non-POSIX
-    mmap/munmap */
+     mmap/munmap */
   cursor->metadata = (struct runtime_events_metadata_header *)
-                      mmap(NULL, cursor->ring_file_size_bytes, PROT_READ,
-                          MAP_SHARED, ring_fd, 0);
-
-  if( cursor->metadata == MAP_FAILED ) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_MAP_FAILURE;
+    mmap(NULL, cursor->ring_file_size_bytes, PROT_READ, MAP_SHARED, ring_fd, 0);
+  if(cursor->metadata == MAP_FAILED) {
+    ret = E_MAP_FAILURE;
+    goto failed;
   }
+
+  ret = E_SUCCESS;
+  /* fallthrough */
+ failed:
+  close(ring_fd);
+  return ret;
 #endif
+}
+
+runtime_events_error caml_runtime_events_create_cursor(
+  const char_os* runtime_events_path, int pid,
+  struct caml_runtime_events_cursor **cursor_res
+) {
+  int ret = E_SUCCESS;
+
+  struct caml_runtime_events_cursor *cursor =
+      caml_stat_alloc_noexc(sizeof(struct caml_runtime_events_cursor));
+  if (cursor == NULL) return E_ALLOC_FAIL;
+  /* zero out all fields, notably the callbacks */
+  memset(cursor, 0, sizeof(*cursor));
+
+  char_os *runtime_events_loc = NULL;
+  ret = format_runtime_events_loc(runtime_events_path, pid, &runtime_events_loc);
+  if (ret != E_SUCCESS) goto failed1;
+
+  ret = cursor_map_ring_file(cursor, runtime_events_loc);
+  if (ret != E_SUCCESS) goto failed2;
 
   cursor->current_positions =
       caml_stat_alloc(cursor->metadata->max_domains * sizeof(uint64_t));
-
   for (int j = 0; j < cursor->metadata->max_domains; j++) {
     cursor->current_positions[j] = 0;
   }
+
   cursor->cursor_open = 1;
   atomic_store(&cursor->cursor_in_poll, 0);
   cursor->next_read_domain = 0;
 
-  cursor->runtime_begin = NULL;
-  cursor->runtime_end = NULL;
-  cursor->runtime_counter = NULL;
-  cursor->alloc = NULL;
-  cursor->lifecycle = NULL;
-  cursor->lost_events = NULL;
-  cursor->user_unit = NULL;
-  cursor->user_int = NULL;
-  cursor->user_span = NULL;
-  cursor->user_custom = NULL;
-
   *cursor_res = cursor;
+  ret = E_SUCCESS;
+  /* fallthrough */
+ failed2:
```

Comment:
Same remarks as for previous functions:
- send all failure cases to the failure section;
- use distinct failure labels for each failure case (even if there is no undo action between a pair of failure labels);
- name each failure label according to the action which has failed;
- only initialize local variables until you know you have succeeded.